pkg/controller/kubelet-config/helpers.go

```diff
@@ -184,6 +184,8 @@ func getManagedKubeletConfigKey(pool *mcfgv1.MachineConfigPool, client mcfgclien
 		if val != "" {
 			return fmt.Sprintf("99-%s-generated-kubelet-%s", pool.Name, val), nil
 		}
+		// the first kubeletconfig is being synced, return its the default MC name
+		return ctrlcommon.GetManagedKey(pool, client, "99", "kubelet", getManagedKubeletConfigKeyDeprecated(pool))
```

Comment:
Ah I see, so what you are saying is that for our condition to trigger, the key being synced is the non-suffixed one (`val == ""` in this case) so we think of it as a new one.

In that case and we apply this fix, if we create KC1 and then KC2, then modify KC1, what is the behaviour expected? Would it be a no-op (since KC1 will cause MC1 but then will be overwritten by MC2)

Reply:
If we create KC1 and then KC2, then modify KC2, we will get no-op. Is this expected?

Reply:
Not only the non-suffixed key got resynced. Print some log `sync kubbeletconfig: key...`, all the existing kubeletconfigs got synced. The new machine config `99-master-generated-kubelet-2` generated and set suffix annotation for max-pod only at the first time it got resynced since it has no `machineconfiguration.openshift.io/mc-name-suffix` annotation at the time when `99-master-generated-kubelet` created. Once it got this annotation, according to the current `getManagedKubeletConfigKey` logic, no machine config was generated in the next time it got resynced.

I can produce bug on both master and worker pools, and not sure why the kubeletconfigs got resynced, and if I pause the pool I can not reproduce the bug.
```
[qiwan@qiwan cluster-dir]$ oc logs -f machine-config-controller-84658b988f-nh2ph
I0127 21:55:17.247968       1 start.go:50] Version: machine-config-daemon-4.6.0-202006240615.p0-1228-g7f7f3a42 (7f7f3a422227e7f4ed0dfd0dc33ea15975d3503d)
I0127 21:55:18.321001       1 leaderelection.go:248] attempting to acquire leader lease openshift-machine-config-operator/machine-config-controller...
I0127 21:57:47.279460       1 leaderelection.go:258] successfully acquired lease openshift-machine-config-operator/machine-config-controller
I0127 21:57:47.292363       1 template_controller.go:137] Re-syncing ControllerConfig due to secret pull-secret change
I0127 21:57:47.381879       1 node_controller.go:152] Starting MachineConfigController-NodeController
I0127 21:57:47.381879       1 template_controller.go:238] Starting MachineConfigController-TemplateController
I0127 21:57:47.381901       1 kubelet_config_controller.go:169] Starting MachineConfigController-KubeletConfigController
I0127 21:57:47.382063       1 kubelet_config_controller.go:172] num of workers:  2
I0127 21:57:47.382167       1 kubelet_config_controller.go:445] sync kubbeletconfig: key: max-podtime; 2022-01-27 21:57:47.382162146 +0000 UTC m=+150.166327122
I0127 21:57:47.382207       1 kubelet_config_controller.go:470] kubeleltc config name: max-pod
I0127 21:57:47.381909       1 container_runtime_config_controller.go:184] Starting MachineConfigController-ContainerRuntimeConfigController
I0127 21:57:47.382293       1 kubelet_config_controller.go:445] sync kubbeletconfig: key: max-pod-2time; 2022-01-27 21:57:47.382290286 +0000 UTC m=+150.166455271
I0127 21:57:47.382306       1 kubelet_config_controller.go:470] kubeleltc config name: max-pod-2
I0127 21:57:47.382901       1 render_controller.go:124] Starting MachineConfigController-RenderController
I0127 21:57:47.485615       1 container_runtime_config_controller.go:802] Applied ImageConfig cluster on MachineConfigPool worker
I0127 21:57:47.563126       1 container_runtime_config_controller.go:802] Applied ImageConfig cluster on MachineConfigPool master
I0127 21:57:47.988463       1 kubelet_config_controller.go:630] Applied KubeletConfig max-pod-2 on MachineConfigPool master
I0127 21:57:48.190772       1 kubelet_config_controller.go:630] Applied KubeletConfig max-pod on MachineConfigPool master
I0127 21:57:48.986469       1 kubelet_config_controller.go:305] Error syncing kubeletconfig max-pod: error deleting degraded kubelet machine config 99-master-generated-kubelet: machineconfigs.machineconfiguration.openshift.io "99-master-generated-kubelet" not found
I0127 21:57:48.986506       1 kubelet_config_controller.go:445] sync kubbeletconfig: key: max-pod-3time; 2022-01-27 21:57:48.986503344 +0000 UTC m=+151.770668306
I0127 21:57:48.986522       1 kubelet_config_controller.go:470] kubeleltc config name: max-pod-3
I0127 21:57:49.991835       1 kubelet_config_controller.go:445] sync kubbeletconfig: key: max-pod-4time; 2022-01-27 21:57:49.991829053 +0000 UTC m=+152.775994010
I0127 21:57:49.991863       1 kubelet_config_controller.go:470] kubeleltc config name: max-pod-4
I0127 21:57:50.591643       1 kubelet_config_controller.go:630] Applied KubeletConfig max-pod-3 on MachineConfigPool master
I0127 21:57:51.390644       1 kubelet_config_controller.go:445] sync kubbeletconfig: key: max-podtime; 2022-01-27 21:57:51.390633019 +0000 UTC m=+154.174797988
I0127 21:57:51.390677       1 kubelet_config_controller.go:470] kubeleltc config name: max-pod
I0127 21:57:51.590763       1 kubelet_config_controller.go:630] Applied KubeletConfig max-pod-4 on MachineConfigPool master
I0127 21:57:52.297084       1 render_controller.go:377] Error syncing machineconfigpool worker: Ignoring controller config generated from machine-config-daemon-4.6.0-202006240615.p0-1227-g9d773078-dirty (my version: machine-config-daemon-4.6.0-202006240615.p0-1228-g7f7f3a42)
I0127 21:57:52.299179       1 render_controller.go:377] Error syncing machineconfigpool master: Ignoring controller config generated from machine-config-daemon-4.6.0-202006240615.p0-1227-g9d773078-dirty (my version: machine-config-daemon-4.6.0-202006240615.p0-1228-g7f7f3a42)


```

Reply:
> If we create KC1 and then KC2, then modify KC2, we will get no-op. Is this expected?

Hmm that sounds wrong. If we create only 1 KC and modify it, does it work?

Reply:
> all the existing kubeletconfigs got synced

That sounds like it could be problematic

> The new machine config 99-master-generated-kubelet-2 generated and set suffix annotation for max-pod only at the first time it got resynced since it has no machineconfiguration.openshift.io/mc-name-suffix annotation at the time when 99-master-generated-kubelet created

Sorry, to clarify, are you saying with this PR, it didn't properly work for the BZ scenario? But it was fine when the pool was paused?

Reply:
> Sorry, to clarify, are you saying with this PR, it didn't properly work for the BZ scenario? But it was fine when the pool was paused?

No, I was saying without this PR the current `func getManagedKubeletConfigKey` generated the new MC name under the BZ scenario. When the pool was paused, I didn't see the kubeletconfigs got resync (may still need to test again to verify this).

> If we create KC1 and then KC2, then modify KC2, we will get no-op. Is this expected?

I was wrong. The kubelet.conf on the node was updated.  After modifying KC2, although no new name `99-worker-generated-kubelet-...`  generated for KC2, the new kubeletconfig was updated by code https://github.com/openshift/machine-config-operator/blob/3a91cb247a103236b5b0c18596e53a3b80f310e6/pkg/controller/kubelet-config/kubelet_config_controller.go#L616-L618

Looking at the master branch code, I think its logic is [func getManagedKubeletConfigKey](https://github.com/openshift/machine-config-operator/blob/3a91cb247a103236b5b0c18596e53a3b80f310e6/pkg/controller/kubelet-config/helpers.go#L148) should always return the same MC name if the kubeletconfig already has a MC, or generate a new MC name if the kubeletconfig has no MC. When resync or edit this kubeletconfig, if its MC exists, try to update the kubeletconfig, if the MC does not exist, create one.
But in the BZ scenario getManagedKubeletConfigKey does not return the right MC name. It generates a new MC name because it forgets to check the case that MC does not have a suffix && len(kcList) > 1. https://github.com/openshift/machine-config-operator/blob/3a91cb247a103236b5b0c18596e53a3b80f310e6/pkg/controller/kubelet-config/helpers.go#L182-L186

With this PR, `oc edit` can update kubeletconfig only when editing the last kubeletconfig.
Without this PR, oc edit` can update kubeletconfig when editing the last kubeletconfig, and the first kubeletconfig(since I think it "wrongly" created a new MC name).

Reply:
> That sounds like it could be problematic

hmm if that's problematic this is just a workaround PR. :thinking:

Reply:
Hmm ok I think I misunderstood initially. If this fixes the BZ issue (and doesn't create a new kubeletconfig) that is a good start :+1: 

> With this PR, oc edit can update kubeletconfig only when editing the last kubeletconfig.

Right, technically, you should still be able to update previous kubeletconfigs (and cause their associated MCs to change). It just won't trigger a new renderedConfig or reboots because the lower priority MCs get overwritten